Unify RBAC checking functions

zmb3 opened the pull request:
Prior to this change, we have several variants of CheckAccessTo*
methods for each resource type that uses RBAC.

This change reduces duplication and the potential for RBAC to behave
differently for different resources by implementing a common
CheckAccess method that can check access to a variety of resources.

This implementation is also sometimes faster than the original
implementation, because it is careful to avoid checking unnecessary
matchers after a role has already denied access.

zmb3 committed:
Unify RBAC checking functions

Prior to this change, we have several variants of CheckAccessTo*
methods for each resource type that uses RBAC.

This change reduces duplication and the potential for RBAC to behave
differently for different resources by implementing a common
CheckAccess method that can check access to a variety of resources.

This implementation is also sometimes faster than the original
implementation, because it is careful to avoid checking unnecessary
matchers after a role has already denied access.

awly commented on lib/services/kube_cluster_resource_wrapper.go:
```diff
@@ -0,0 +1,56 @@
+	"github.com/gravitational/teleport/api/types"
+)
+
+// KubeClusterResourceWrapper stubs out a minimal implementation of
+// services.AccessCheckable for *types.KubernetesCluster.
+// This allows us to use common RBAC checks for Kubernetes Clusters
+// and other resources.
```
The only thing really missing from `KubernetesCluster` is namespace, right?
How about implementing the missing methods (except `GetMetadata`) on `KubernetesCluster` itself, and embedding it in this wrapper. The wrapper can basically be:
```
type KubeClusterResourceWrapper {
  *types.KubernetesCluster
  Namespace string
}
func (k KubeClusterResourceWrapper) GetMetadata() types.Metadata {
	return types.Metadata{
		Name:      k.GetName(),
		Namespace: types.ProcessNamespace(k.Namespace),
	}
}
```

zmb3 replied:
That works, I was afraid of making `KubernetesCluster` look too much like a `types.Resource` and making this a much larger refactor, but I like this suggestion.

zmb3 replied:
I've applied your suggestion to just take the namespace instead of holding a reference to the server, but am still not sold on modifying the `api/types` package. I don't want to confuse people consuming the API by making this look like any other resource when it's actually handled quite differently.

What do you think @awly? Good as is or would you really want to see this change?

awly commented on lib/services/role.go:
```diff
@@ -1887,69 +1616,155 @@ func (m *DatabaseNameMatcher) String() string {
-// The checker always checks the server namespace, other matchers are supplied
-// by the caller.
-func (set RoleSet) CheckAccessToDatabase(server types.Database, mfa AccessMFAParams, matchers ...RoleMatcher) error {
+// Note: logging in this function only happens in debug mode. This is because
+// adding logging to this function (which is called on every resource returned
+// by the backend) can slow down this function by 50x for large clusters!
```
nit: move this comment inside the function, before you create the `debugf` wrapper, I missed it on first pass.
Also, explain why your wrapper is much faster than `log.Debugf`. Also also, consider fixing `log.Debugf` upstream to do what your helper does.

zmb3 replied:
The helper is just to put the if statement in one place rather check the log level before every call to `log.Debugf`.

The comment and the associated code is not new, @russjones wrote it years ago. I'm not sure whether its necessary or if log.Debugf` is automatically a no-op when the level is not debug, but the comment sounds like this logic was created due to major performance issues, so I wanted to preserve the existing behavior.

zmb3 replied:
Moved comment inside, and fixed another occurrence of the same pattern in this file.

awly commented on lib/services/role.go:
```diff
@@ -1887,69 +1616,155 @@ func (m *DatabaseNameMatcher) String() string {
+	GetAllLabels() map[string]string
+}
+
+// CheckAccess checks if this role set has access to a particular resource.
+// It attempts to match labels if roleLabelsFn is not nil.
```
Expand this comment a bit more, and maybe add an example.
The usage is somewhat clever, especially for people not familiar with method values.
Explain what usually gets passed as `roleLabelsFn` and `matchers` and what the behavior is with no `matchers`.

Alternatively, you can preserve something similar to the old API, but make them all be thin wrappers around `set.checkAccess`. Type-specific wrappers in one place may be slightly less error prone (errors like using the wrong set of labels, forgetting to add a `RoleMatcher` for SSH logins, etc).

zmb3 replied:
I saw this comment coming :-) I agree its a bit clever and maybe a bit too easy to make a mistake and thought that strongly typed wrappers would be a good solution.

I do like that this approach doesn't require new methods in the public API every time we add a new resource, though. Teleport has a ton of that and it makes adding new things take longer than it should.

zmb3 replied:
Alternative idea - the method could accept a bool for `matchLabels` instead of a func, and if it is set, then we can decide how to fetch the labels internally based on the type and/or kind of `r`. 

Makes it harder to mess up and look at the wrong labels that way.

codingllama replied:
LGTM.

zmb3 commented on lib/auth/auth.go:
```diff
@@ -3053,7 +3063,11 @@ func (a *Server) isMFARequired(ctx context.Context, checker services.AccessCheck
+		noMFAAccessErr = checker.CheckAccess(
+			db,
+			services.AccessMFAParams{AlwaysRequired: false, Verified: false},
+			nil, // we do not check labels for databases
```
This is the only place we didn't. Other uses of `CheckAccessToDatabase` all passed a `DatabaseLabelMatcher`, so we're just preserving the original behavior here.

awly replied:
@r0mant could you confirm that this is intentional?

zmb3 replied:
I believe this to be intentional. This code is attempting to determine if MFA is required by running the check (without label matching) and seeing if the error that comes back is `ErrSessionMFARequired`.

If we do match labels, and the RBAC check fails due to labels, then we'll return access denied (before even considering MFA) instead of `ErrSessionMFARequired` and this function will report that MFA is not required.

That said, the same checks for MFA being enabled on nodes and k8s clusters _do_ match labels.

r0mant replied:
I don't know, I didn't write this particular piece code but it seems like an omission to me. The labels will be checked by the database agent anyway but we could do it here too for consistency. Seems like if the labels don't match, there's no point in checking MFA either.

zmb3 replied:
Ok cool, I will remove the argument entirely and always check labels. Seems like what we'd want anyway since labels are core to our whole RBAC system.

zmb3 commented on lib/services/role.go:
```diff
@@ -1887,69 +1616,155 @@ func (m *DatabaseNameMatcher) String() string {
+// LoginMatcher is a RoleMatcher that checks whether the role's logins
+// match the specified condition.
+type LoginMatcher struct {
+	login *string
```
A relic of of chasing down some performance regressions and trying to reduce allocations. Will fix.

r0mant committed:
Set flush interval when forwarding application http requests (#8359)

Signed-off-by: Roman Tkachenko <roman@gravitational.com>

Co-authored-by: Stefan Sedich <stefan.sedich@gmail.com>

zmb3 committed:
First pass at addressing review feedback

russjones committed:
[auto] Update AMI IDs for 7.2.0

andrejtokarcik committed:
Migrate DynamoDB events to store fields as map type (#8292)

codingllama commented on lib/services/role_test.go:
```diff
@@ -3035,9 +3049,14 @@ func BenchmarkCheckAccessToServer(b *testing.B) {
-				}
+				// note: we don't check the error here because this benchmark
+				// is testing the performance of failed RBAC checks
+				set.CheckAccess(
```
Suggestion:

```suggestion
				_ = set.CheckAccess(
```

codingllama commented on lib/web/ui/usercontext.go:
```diff
@@ -115,8 +115,7 @@ func getLogins(roleSet services.RoleSet) []string {
 	for _, login := range allowed {
-		loginMatch, _ := services.MatchLogin(denied, login)
-		if !loginMatch {
+		if isDenied := apiutils.SliceContainsStr(denied, login); !isDenied {
```
The code looked more robust before (MatchLogin vs Contains check), why did you elect to change it?

zmb3 replied:
Because `MatchLogin` doesn't exist anymore - it's been replaced with `services.NewLoginMatcher`. The matchers have slightly different semantics, this code is really just filtering out denied logins and doesn't need to conform to the matcher interface.

codingllama replied:
I knew I was missing something, thanks.

zmb3 commented on lib/services/role.go:
```diff
@@ -1887,69 +1616,155 @@ func (m *DatabaseNameMatcher) String() string {
-		// matchers return true.
-		if matchNamespace {
-			match, err := RoleMatchers(matchers).MatchAll(role, Allow)
+		matchNamespace, namespaceMessage := MatchNamespace(role.GetNamespaces(Allow), namespace)
```
I'll give it a try and see how it looks. In addition to getting the allow/deny labels, the matchers use `MatchAll` for for allow rules, and `MatchAny` for deny rules.

codingllama replied:
Yes, a few more things need to be flipped around. Let me know how that goes.

zmb3 replied:
It looks to me like this isn't a meaningful improvement. You need to pass quite a bit of context to the helper, and there's a lot of special cases (debugging is more verbose for allow rules than deny rules, you have to invert the conditions on all the matchers, etc)

I'm not going to proceed with this change right now. We can always revisit later.

codingllama replied:
Fair enough, thanks for giving it a shot.

zmb3 committed:
Address more review feedback

zmb3 committed:
Fix debug logging

Use IsLevelEnabled, which checks that the log level is *at least* debug.

zmb3 committed:
Determine how to get labels from the role automatically

This makes it harder to mess up and look at the wrong labels when
checking the RBAC rules for a resource.

codingllama commented on lib/auth/auth.go:
```diff
@@ -3000,7 +3000,12 @@ func (a *Server) isMFARequired(ctx context.Context, checker services.AccessCheck
+			err := checker.CheckAccess(
+				n,
+				services.AccessMFAParams{},
+				true,
```
Annotate booleans so their meaning is clear?

```suggestion
				true /* matchLabels */,
```

rosstimothy replied:
or refactor to remove `bool` all together?

codingllama commented on lib/services/role.go:
```diff
@@ -1887,69 +1600,175 @@ func (m *DatabaseNameMatcher) String() string {
+	namespace := types.ProcessNamespace(r.GetMetadata().Namespace)
+	allLabels := r.GetAllLabels()
+
+	type roleLabelGetterFn func(types.Role, types.RoleConditionType) types.Labels
+	var getRoleLabels roleLabelGetterFn
```
Maybe:

```suggestion
	var getRoleLabels func(types.Role, types.RoleConditionType) types.Labels
```

codingllama reviewed (approved):
LGTM.

There a few open comments, but nothing too serious, so leaving it pre-approved.

zmb3 committed:
Remove unnecessary type def

rosstimothy commented on lib/services/role.go:
```diff
@@ -713,8 +713,8 @@ type AccessChecker interface {
-	// CheckAccessToServer checks access to server.
-	CheckAccessToServer(login string, server types.Server, mfa AccessMFAParams) error
+	// CheckAccess checks access to the specified resource.
+	CheckAccess(r AccessCheckable, mfa AccessMFAParams, matchLabels bool, matchers ...RoleMatcher) error
```
Would be nice to replace the `bool` with an `enum` or maybe split this into `CheckAccessWithLabels` and `CheckAccessWithoutLabels` that way the intent is a bit more clear to callers. Although this interface alone has several other functions with boolean parameters.

r0mant replied:
I agree. It also feels weird that matchLabels is separate from other matches for some reason. Can it just be one of the matchers?

zmb3 replied:
Will make an enum for this. We might be able to remove it altogether with your confirmation on https://github.com/gravitational/teleport/pull/8407#discussion_r718799605

codingllama committed:
Test Webauthn global disable flag (#8393)

awly commented on lib/services/role.go:
```diff
@@ -1299,16 +1279,15 @@ func (set RoleSet) GetLoginsForTTL(ttl time.Duration) (logins []string, matchedT
-// Note, logging in this function only happens in debug mode, this is because
-// adding logging to this function (which is called on every server returned
-// by GetRemoteClusters) can slow down this function by 50x for large clusters!
 func (set RoleSet) CheckAccessToRemoteCluster(rc types.RemoteCluster) error {
```
Any way to roll this into `CheckAccess`?

zmb3 replied:
I looked briefly and some of the extra backwards compatibility logic in there made me hesitant to change it. I will take a closer look and follow up with a subsequent PR if I can find a way to remove this.

awly reviewed (approved):
LGTM

Check with @r0mant about the one place where labels are not matched.
If labels should be matched there (or if matching doesn't hurt), remove the bool `matchLabels` parameter entirely.

zmb3 committed:
Merge branch 'master' into zmb3/unify-rbac2

r0mant commented on lib/services/role.go:
```diff
@@ -1887,69 +1600,174 @@ func (m *DatabaseNameMatcher) String() string {
+	allLabels := r.GetAllLabels()
+
+	var getRoleLabels func(types.Role, types.RoleConditionType) types.Labels
+	if matchLabels {
+		switch r.GetKind() {
+		case types.KindDatabase:
+			getRoleLabels = types.Role.GetDatabaseLabels
+		case types.KindApp:
+			getRoleLabels = types.Role.GetAppLabels
+		case types.KindNode:
+			getRoleLabels = types.Role.GetNodeLabels
+		case kindKubeClusterRBAC:
+			getRoleLabels = types.Role.GetKubernetesLabels
+		default:
+			return trace.BadParameter("cannot match labels for kind %v", r.GetKind())
+		}
+	}
```
If we have callers pass their respective label matchers, this won't be needed.

zmb3 replied:
That's how it was originally. But it was then possible to call incorrectly.  For example, you could make a mistake like the following and perform checks against a role's database labels for an app.

```
var app types.App
CheckAccess(app, ..., types.Role.GetDatabaseLabels)
```

r0mant replied:
I mean, you can do the same now as well if you pass other incorrect matchers? Not sure how label matcher is different. It's just very weird from the function contract perspective that we seem to be unifying the checkers and matchers but labels are for some reason treated differently.

r0mant commented on lib/services/role.go:
```diff
@@ -1887,69 +1600,174 @@ func (m *DatabaseNameMatcher) String() string {
+	return &loginMatcher{login: login}
+}
+
+func (l *loginMatcher) Match(role types.Role, typ types.RoleConditionType) (bool, error) {
```
Nit: Missing godoc.

codingllama committed:
Implement attestation for Webauthn (#8392)

Attestation allows Teleport admins a manner of control over the devices used for
MFA.

Attestation support for Webauthn works similarly to U2F attestation, but
provides both allow and deny lists for device CAs. The device's attestation
certificate must clear both lists before registration is complete.

Previously-registered devices are not affected (similarly to U2F).

* Better error messages on webauthn/httpserver
* Make mocku2f attestation cert public
* Implement attestation for Webauthn

r0mant commented on lib/services/role.go:
```diff
@@ -1887,69 +1600,174 @@ func (m *DatabaseNameMatcher) String() string {
+	GetAllLabels() map[string]string
+}
+
+func rbacDebugLogger() (debugEnabled bool, debugf func(format string, args ...interface{})) {
```
Nit: Explain with a godoc the purpose of this method.

zmb3 replied:
There's an explanation at the call site, which is where other reviewers requested it, but I'm happy to move it.

r0mant replied:
Yeah I saw it, I think it's worth explaining it here as well (e.g. so somebody wanting to know what this does doesn't have to search for where it's called).

r0mant commented on lib/services/role.go:
```diff
@@ -1887,69 +1600,174 @@ func (m *DatabaseNameMatcher) String() string {
+}
+
+// AccessCheckable is the subset of types.Resource required for the RBAC checks.
+type AccessCheckable interface {
```
Why not just use `types.Resource`? It's our base interface for all resources.

zmb3 replied:
Because:

1) it's huge - there's way more methods in there than we actually need, making it harder to mock out and test
2) `types.KubernetesCluster` is not a `types.Resource`, so the wrapper necessary to make that work would require us to stub out even more methods we never plan to call.
2) We actually need `types.ResourceWithLabels`, not `types.Resource`, but unfortunately that interface also embeds `Origin() string` which not all resources implement, requiring further updates to all resources.

r0mant replied:
For mocking you can just embed `types.Resource` into your test struct and only implement the methods you need and with my suggestion below KubeCluster will also be a resource. But it's no big deal, specialized interfaces are fine.

r0mant commented on lib/services/kube_cluster_resource_wrapper.go:
```diff
@@ -0,0 +1,58 @@
+
+const kindKubeClusterRBAC = "kube_cluster"
+
+func NewKubernetesClusterRBAC(namespace string, kube *types.KubernetesCluster) KubernetesClusterRBAC {
```
The proper way would be to make a proper `KubernetesCluster` resource. This would also pave the way (somewhat) for separating it into a separate resource in future (similar to apps and databases).

zmb3 replied:
Correct. Andrew advised against going down that path for now and revisiting in the future as it's a larger refactor and the goal of this is just to pave the way for desktop access, not rewrite Kubernetes access.

r0mant replied:
This feels quite hacky TBH. I would spend a little extra time and add a proper KubernetesCluster resource. You don't need to refactor a whole lot, just define a resource and a conversion method from old struct (which is a part of the Server) to the resource. I did the same for Apps.

zmb3 replied:
Defined the new resource and conversion method. Let me know what you think.

stevenGravy committed:
correct app name example (#8422)

was mixing jira and jenkins

zmb3 committed:
Address review feedback

- remove flag to match labels - prefer to always match
- make a new resource for KubernetesClusterV3 and a conversion func

Note: plumbing the KubernetesClusterV3 resource through existing protos
and deprecating the original resource is out of scope for this change.

zmb3 committed:
Merge branch 'master' into zmb3/unify-rbac2

r0mant commented on lib/services/role.go:
```diff
@@ -1887,69 +1600,169 @@ func (m *DatabaseNameMatcher) String() string {
+	namespace := types.ProcessNamespace(r.GetMetadata().Namespace)
+	allLabels := r.GetAllLabels()
+
+	var getRoleLabels func(types.Role, types.RoleConditionType) types.Labels
+	switch r.GetKind() {
+	case types.KindDatabase:
+		getRoleLabels = types.Role.GetDatabaseLabels
+	case types.KindApp:
+		getRoleLabels = types.Role.GetAppLabels
+	case types.KindNode:
+		getRoleLabels = types.Role.GetNodeLabels
+	case types.KindKubernetesCluster:
+		getRoleLabels = types.Role.GetKubernetesLabels
+	default:
+		return trace.BadParameter("cannot match labels for kind %v", r.GetKind())
+	}
```
What do you think about encapsulating this as a part of `types.Role`? E.g. a role can have a method

```go
GetLabelsFor(AccessCheckable, types.RoleConditionType) (types.Labels, error)
```

Would make this function a little cleaner.

zmb3 replied:
Sure, but that moves `AccessCheckable` into types. We could also do `GetLabelsForKind(kind string, ..)`

r0mant commented on api/types/types.proto:
```diff
@@ -2075,6 +2077,29 @@ message KubernetesCluster {
+    KubernetesClusterSpecV3 Spec = 5 [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "spec" ];
+}
+
+message KubernetesClusterSpecV3 {
```
Nit: Add godoc.

zmb3 committed:
Add missing godoc

zmb3 committed:
Merge branch 'master' into zmb3/unify-rbac2
